[FEATURE] Elastic 6.x support

bmoex opened the pull request:
- Removed multiple documents from indexation
- Added specific deletion from same index
- Bugfixed the Generator loop when loop was not finished but has no results in the 50 items
- Added a new Functional test for deleteDocuments
- Added typolink to indexation
- Seperated form & list plugin for flexible usage
- Added frontend access lookup for searching in hidden/public pages

Commit:
Merge pull request #165 from Codappix/release/0.0.7

Release/0.0.7

bmoex committed:
[TASK] Update elastica version pointer

bmoex committed:
[TASK] Add docprops to classes

+ Updated version pointer
+ Replaces deprecated ArrayUtility for Core utility

bmoex committed:
[TASK] Remove deprecated _all field query for Elastic 6.x

See: https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-all-field.html

bmoex committed:
[TASK] Add typolink to index

bmoex committed:
[TASK] Add both form as list plugin seperate

bmoex committed:
[TASK] Corectly index access group for query content

bmoex committed:
[TASK] Add transparent background color for extension icon

bmoex committed:
[!!!][FEATURE] Remap specific document type's to ES6 identifiers

Multiple types are not allowed any more in ES6. Refactored to use document type and custom
identifier to still create different 'types'.

See: https://www.elastic.co/guide/en/elasticsearch/reference/6.x/removal-of-types.html

bmoex committed:
[TASK] Add opening for possible user access lookup

bmoex committed:
[FEATURE] Add Frontend User access query for secure search on elastic index

bmoex committed:
[BUGFIX] Make sure the while loop is not closed when filterRecordsByRootLineBlacklist() returns no results

bmoex committed:
[BUGFIX] Inherit access groups from parent pages when extended to subpages

bmoex committed:
[TASK] Finetune query for document_type instead of regex

bmoex committed:
[TASK] Process feedback and fix unit-tests

bmoex committed:
[BUGFIX] Use relative location of Fixtures

bmoex committed:
[TASK] Process own feedback based on tests

bmoex committed:
[TASK] Add functional test for deleteDocuments functionality

DanielSiepmann commented:
Thanks for this amazing work @bmoex. Can you check the broken tests and fix them?

I need to check when I've time to review all these changes.

bmoex committed:
[BUGFIX] Remove _all from unit tests

bmoex committed:
[FEATURE] Added flush command as indexation is now into one

bmoex committed:
[TASK] Make sure CGL is correct

bmoex committed:
[TASK] Try composer update for travis build

bmoex committed:
[TASK] Enable elasticsearch version 6 for travis

bmoex committed:
[TASK] Remove support for TYPO3 7.6

DanielSiepmann committed:
TASK: Remove code base for TYPO3 7.x

Also add documentation with breaking change.

DanielSiepmann committed:
TASK: Prevent php warning for stdClass constant

DanielSiepmann committed:
FEATURE: Always accept comma separated list of identifiers on CLI

* Streamline all commands to always accept a comma separated list of
  identifiers.
* Adjust phpdoc to reflect this feature and provide help on CLI.
* Refactor code to move recurring logic to own method.
* Provide tests for new feature.
* Add documentation for feature.

DanielSiepmann commented on Classes/Configuration/ConfigurationContainer.php:
```diff
@@ -39,14 +40,14 @@ class ConfigurationContainer implements ConfigurationContainerInterface
         $this->settings = $configurationManager->getConfiguration(
             ConfigurationManagerInterface::CONFIGURATION_TYPE_SETTINGS,
             'SearchCore',
-            'search'
```
I'm not sure about that change yet. Why is this necessary?

DanielSiepmann commented on Classes/Connection/ConnectionInterface.php:
```diff
@@ -20,6 +21,8 @@
  * 02110-1301, USA.
  */
 
+use Elastica\Query;
```
The interface should not know anything about Elastica.
This is up to the concrete implementation.

Wouldn't it be possible to re-use the existing `SearchRequestInterface`?

DanielSiepmann commented on Classes/Connection/Elasticsearch/DocumentFactory.php:
```diff
@@ -44,17 +45,21 @@ public function injectLogger(\TYPO3\CMS\Core\Log\LogManager $logManager)
         }
 
         $identifier = $document['search_identifier'];
-        unset($document['search_identifier']);
```
Why was that line removed? This way the information is added twice to the document which should not be necessary.

DanielSiepmann commented on Classes/Connection/Elasticsearch/IndexFactory.php:
```diff
@@ -65,15 +66,19 @@ public function __construct(ConfigurationContainerInterface $configuration)
+    public function getIndexName(): string
     {
-        return $this->configuration->get('connections.elasticsearch.index');
+        return (string)$this->configuration->get('connections.elasticsearch.index');
```
The casting is not necessary. We do not use strict declarations. Adding `: string` will cast the return value to string already.

DanielSiepmann commented on Classes/Domain/Index/AbstractIndexer.php:
```diff
@@ -23,6 +24,7 @@
 use Codappix\SearchCore\Configuration\ConfigurationContainerInterface;
 use Codappix\SearchCore\Configuration\InvalidArgumentException;
 use Codappix\SearchCore\Connection\ConnectionInterface;
+use Elastica\Query;
```
The abstract indexer should not know anything about Elastica. It's completely decoupled from concrete implementations like Elasticsearch.

DanielSiepmann commented on Classes/Domain/Index/TcaIndexer/PagesIndexer.php:
```diff
@@ -123,17 +135,72 @@ protected function fetchContentForPage(int $uid) : array
+    protected function fetchAccess(int $uid, array $pageAccess): array
+    {
+        try {
+            $objectManager = GeneralUtility::makeInstance(ObjectManager::class);
```
There should be no hardcoded dependency, when not necessary. Instead use Dependency Injection to get an instance of the ObjectManager.

DanielSiepmann commented on Classes/Domain/Index/TcaIndexer/RelationResolver.php:
```diff
@@ -101,15 +126,22 @@ protected function getUtilityForMode() : string
             $record['uid']
-        ) ?? '';
+        );
+        return $value ? (string)$value : '';
```
Why is this change necessary? PHP will cast to string due to return type hint. The fallback to an empty string was also given before.

bmoex replied:
$utility::getProcessedValueExtra returned ```int``` type

DanielSiepmann commented on Classes/Domain/Index/TcaIndexer/TcaTableService.php:
```diff
@@ -132,27 +148,65 @@ public function getRecord(int $identifier) : array
     public function prepareRecord(array &$record)
     {
-        if (isset($record['uid']) && !isset($record['search_identifier'])) {
-            $record['search_identifier'] = $record['uid'];
```
Why do we no longer provide the uid, but a combined identifier to the search?

Shouldn't each document type have an own index?

bmoex replied:
In my current indexation its still 1 index, but have multiple internal types. Therefor a combined string identifier is required.

DanielSiepmann commented on Classes/Domain/Index/TcaIndexer/TcaTableService.php:
```diff
@@ -132,27 +148,65 @@ public function getRecord(int $identifier) : array
+            $record['search_access'] = !empty($groups) ? $groups : [0];
+        }
+
+        if (!isset($record['search_page_typolink'])) {
```
If the link is not always to a page, I would it only call `search_typolink`.

DanielSiepmann replied:
I would try to use https://github.com/TYPO3/TYPO3.CMS/blob/TYPO3_8-7/typo3/sysext/core/Classes/LinkHandling/LinkService.php#L153 for all of these. This way we only have configuration and do not rewrite existing logic. Also one can register custom link handler, which is not supported by the current approach, but would be once we use the public API.

DanielSiepmann replied:
We should add an entry to the changelog documentation about the new feature.

DanielSiepmann replied:
Why do we need this feature? We built links within fluid templates where we had the result of elasticsearch. There we would check for a type and would know how to render the link. There we also have all our provided settings as an integrator and could provide page uids for custom records.

bmoex replied:
I will remove this functionality. This could easily be handled in a view as well.

DanielSiepmann commented on Classes/Domain/Index/TcaIndexer/TcaTableService.php:
```diff
@@ -132,27 +148,65 @@ public function getRecord(int $identifier) : array
+                    $record['search_page_typolink'] = 't3://file?uid=' . $record['uid'];
+                    break;
+                default:
+                    $record['search_page_typolink'] = 't3://page?uid=' . $record['pid'];
```
This will not work for records within sys folders. Still a link would be generated, but this wont work.

DanielSiepmann commented on Classes/Utility/ArrayUtility.php:
```diff
@@ -0,0 +1,34 @@
+ * Utility: Array
+ * @package Codappix\SearchCore\Utility
+ */
+class ArrayUtility
```
Why not extend the core utility? This way one doesn't have to use both classes. We just extend the original feature set.

bmoex replied:
I don't like to extend utility classes with static functions. But will change :-)

DanielSiepmann commented on Classes/Domain/Model/SearchRequest.php:
```diff
@@ -117,36 +134,47 @@ public function addFacet(FacetRequestInterface $facet)
         $this->searchService = $searchService;
     }
 
-    // Extbase QueryInterface
```
This is not only about the execute method, but also the following methods.

That's the reason this was not part of the phpdoc.

DanielSiepmann commented on Classes/Domain/Model/SearchRequest.php:
```diff
@@ -117,36 +134,47 @@ public function addFacet(FacetRequestInterface $facet)
+     * Current implementation covers only paginate widget support.
+     *
+     * @param bool $returnRawQueryResult
+     * @return array|\Codappix\SearchCore\Connection\SearchResultInterface|\TYPO3\CMS\Extbase\Persistence\QueryResultInterface
```
We only return our own SearchResultInterface, there is no way we could return an array or extbase query result.

DanielSiepmann commented on Classes/Domain/Model/SearchRequest.php:
```diff
@@ -156,140 +184,270 @@ public function execute($returnRawQueryResult = false)
     }
 
+    /**
+     * @return \TYPO3\CMS\Extbase\Persistence\Generic\Qom\SourceInterface|void
```
There is no way that anything could be returned.

DanielSiepmann commented on Classes/Domain/Model/SearchResult.php:
```diff
@@ -20,9 +21,7 @@
  * 02110-1301, USA.
  */
 
-use Codappix\SearchCore\Connection\ResultItemInterface;
```
These interfaces are still used within return type annotations. IDEs should not be able to resolve those when the use statements are removed.

DanielSiepmann commented on Classes/Domain/Model/SearchResult.php:
```diff
@@ -62,66 +66,94 @@ public function __construct(SearchResultInterface $originalSearchResult, array $
-            $this->results[] = new ResultItem($item['data'], $item['type']);
+        if ($this->results === null) {
+            foreach ($this->resultItems as $item) {
+                $this->results[] = new ResultItem($item['data'], $item['type']);
```
This should produce a warning, as `$this->results` is `null` and no longer an array.

DanielSiepmann commented on Classes/Domain/Search/CachedSearchService.php:
```diff
@@ -0,0 +1,57 @@
+<?php
+
+namespace Codappix\SearchCore\Domain\Search;
+
```
Please always add the license header to files. This is mandatory for the used license.

DanielSiepmann commented on Classes/Domain/Search/CachedSearchService.php:
```diff
@@ -0,0 +1,57 @@
+ * Service: Cached Search
+ * @package Codappix\SearchCore\Domain\Search
+ */
+class CachedSearchService implements SingletonInterface
```
As we now have two classes doing the same, I would like to have an interface. This way one can decide which implementation to use and can also decide to use a custom implementation.
The usages, e.g. controller, should than require the interface.

Right now we use one or the other without any interface.

This class will just work as a caching proxy which is nice. This might be the default registered implementation.

DanielSiepmann commented on Classes/Domain/Search/QueryFactory.php:
```diff
@@ -183,84 +230,114 @@ protected function addFields(SearchRequestInterface $searchRequest, array &$quer
+                    'value' => $value,
+                    'query' => &$query,
+                ];
+                GeneralUtility::callUserFunction($config['userFunc'], $parameters, $this);
```
I like the idea, but I would use an interface, like for DataProcessing, and provide each parameter.
So one can configure a class instead of a userfunction. This class needs to implement the interface with one method. The interface defined the three arguments provided.

This way it should be easier for developers within IDEs and editors. Also breaking changes can be provided through the interface, making it more clear what has changed.

The developer would not retrieve an "array with something" but clean defined arguments. And therefore we should also force him to return the modified query. As otherwise we force him to define `$query` as reference.

DanielSiepmann commented on Classes/Integration/Form/Finisher/DataHandlerFinisher.php:
```diff
@@ -46,10 +47,15 @@ class DataHandlerFinisher extends AbstractFinisher
     ];
 
+    /**
+     * @return null|string|void
```
I can not see that anything is returned at all.

DanielSiepmann commented on Configuration/TCA/Overrides/tt_content.php:
```diff
@@ -1,3 +1,23 @@
+use TYPO3\CMS\Extbase\Utility\ExtensionUtility;
+
+call_user_func(function ($extension, $table) {
+    $plugin = ExtensionUtility::registerPlugin(
```
Is this a new feature with 9.5? In 8.7 nothing is returned from this method call.

Also, if the value is hardcoded afterwards anyway, we do not need this?

bmoex replied:
Sorry, this is something I always do for readability.

DanielSiepmann commented on Configuration/TCA/Overrides/tt_content.php:
```diff
@@ -1,3 +1,23 @@
+            'Codappix.' . $extension,
+            'Results',
+            'LLL:EXT:search_core/Resources/Private/Language/locallang_be.xlf:plugin.results.title',
+            'EXT:search_core/Resources/Public/Icons/PluginResults.svg'
```
Shouldn't we use icon identifiers instead?

DanielSiepmann commented on Configuration/TypoScript/constants.typoscript:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    settings {
+        searchOverview =
```
What's this setting for? I could not find anything, either in code nor docs.

DanielSiepmann commented on Configuration/TypoScript/setup.typoscript:
```diff
@@ -0,0 +1,59 @@
+    view {
+        # This sets the plugin namespace of all your plugins inside the extension to "tx_searchcore",
+        # making it possible to directly access search_core parameters in your controller:
+        pluginNamespace = tx_searchcore
```
If we use the same namespace for all plugins, why do we have two plugins at all?

If it's only for integrator, he can add a flexform input with a selector or some custom setting via TypoScript to render a different section or something else.

DanielSiepmann commented on Resources/Private/Language/locallang_be.xlf:
```diff
@@ -0,0 +1,20 @@
+                <source>Search Core: Form</source>
+            </trans-unit>
+            <trans-unit id="plugin.form.description">
+                <source>Just a search form to list page</source>
```
What's meant by `to list page`?

DanielSiepmann commented on Resources/Private/Layouts/Default.html:
```diff
@@ -0,0 +1,6 @@
+<html xmlns:f="http://xsd.helmut-hummel.de/ns/TYPO3/CMS/Fluid/ViewHelpers"
```
I would keep this the same through all templates, so use `typo3.org` instead.

DanielSiepmann committed:
TASK: Fix phpdocs and type hints

DanielSiepmann committed:
TASK: Add missing license comments in PHP files

DanielSiepmann committed:
TASK: Refactor code to use guards and decrease indentation

DanielSiepmann committed:
TASK: Use same order within all files for plugins

DanielSiepmann commented on Classes/Hook/Filter/FrontendUserAccessFilter.php:
```diff
@@ -0,0 +1,82 @@
+<?php
+
+namespace Codappix\SearchCore\Hook\Filter;
```
Why did you choose "hook" as a namespace?

DanielSiepmann committed:
TASK: Streamline phpdoc

* Do not add duplicate information from PHP to phpdoc.
* Do not add useless comments like "this is a constructor of class x".

bmoex committed:
[TASK] Finetune so interface is not reliant for Elastica library

bmoex committed:
[TASK] Remove duplicate search_identifier in index

bmoex committed:
[TASK] Remove double variables

bmoex committed:
[BUGFIX] Make sure the while loop is not closed

When filterRecordsByRootLineBlacklist() returns no results, loop should just ask
the next items based on iteration

bmoex committed:
[BUGFIX] Make sure boolean is returned for strict return type

When configured by comma seperate list, it should return true

bmoex committed:
[TASK] Use generic namespace for fluid

bmoex committed:
[TASK] Added link generation based on fluid templates

Removed search_page_typolink parameter from index

bmoex committed:
[TASK] Add documentation for frontend user access feature

bmoex committed:
[TASK] Add objectmanager via constructor DI injection

bmoex committed:
[TASK] Revert to only one plugin based on feedback

bmoex committed:
[FEATURE] Use interface for SearchService

Added cached proxy service for data interaction

bmoex committed:
[TASK] Update unit tests for renaming controller initialize

bmoex committed:
[TASK] Remove unused constant

bmoex committed:
[TASK] Use filter interface for custom added filter

bmoex committed:
[TASK] Remove unneeded pluginNamespace

bmoex committed:
[TASK] Finetune QueryFactoryTest based on changes

bmoex committed:
[TASK] Apply CGL changes

bmoex committed:
[TASK] Add a little description howto extend configuration

DanielSiepmann committed:
Merge remote-tracking branch 'origin/develop' into feature/elastic-upgrade

Conflicts:
    Classes/Configuration/ConfigurationContainer.php

DanielSiepmann committed:
BUGFIX: Keep default action non cacheable

Otherwise user input is saved and re displayed to foreign visitors.

If someone is in need for an cached version, he can create a plugin
inside his sitepackage with a different plugin signature but same
action.

If that does not work, we might add another action for caching.

Also keep old action name to be non breaking.

DanielSiepmann committed:
BUGFIX: Keep necessary type hint for extbase

DanielSiepmann committed:
TASK: Remove custom filter for fe group filtering

As this didn't work on first use. Also no tests are available for this
feature yet.

As we remove the whole feature in a single commit, we can revert this
commit later. This way we can upgrade Elasticsearch and provide this
feature later on.

DanielSiepmann committed:
TASK: Update elasticsearch even further

Use latest stable 6.x releases, for elasticsearch and used library.

DanielSiepmann committed:
TASK: Fix / improve some comments

* Add missing throws annotation.
* Remove useless comments.
* Import namespaces used in comments.

DanielSiepmann committed:
TASK: Refactor type handling for elasticsearch

With V6 of elasticsearch the meaning / handling of index and type has
changed, and it will change further in the future.

We therefore move code to own classes and adjust / refactor calling
code.

DanielSiepmann committed:
TASK: Shorten code

* Do not use variable, instead provide result to method call.
* Do not provide public method which is not part of API. Otherwise one
  might use it in custom PHP code and switch connection later on. Then the
  method call will no longer work.

DanielSiepmann committed:
TASK: Refactor code into common flow

As we have the same flow in some more places, we refactor this part to
follow this conventions.

DanielSiepmann committed:
TASK: Cleanup TypoScript

Do not provide unnecessary Constants or Setup.

Fluid paths are there out of the box and integrators have all freedom
without the provided constants.

DanielSiepmann committed:
TASK: Improve documentation

* Add some missing documentation.
* Adjust some documentation due to breaking changes.
* Adjust structure of documentation for easier changelog rendering and
  browsing.

DanielSiepmann committed:
TASK: Use new SearchRequestInterface everywhere

Otherwise the interface is no big benefit. Also it might crash under
some circumstances.

DanielSiepmann committed:
TASK: Fix unit tests

DanielSiepmann committed:
TASK: Use php 7 anonymous function call

DanielSiepmann committed:
TASK: Remove unused file / php class

DanielSiepmann committed:
TASK: Fix license and php doc

DanielSiepmann committed:
TASK: rephrase indice to index.

DanielSiepmann committed:
TASK: Finish deletion of index and documents feature

We replace the "flush" and "delete" by "delete" and "deletedocuments"
logic. This makes it more obvious what will happen, without reading the
docs.

Also we kept the logic to always provide the index name, as
we will need them in the future. Due to elasticsearch v6 changes no
types are allowed in the same index in the future. Therefore we need to
make it possible to use different indexes in the future, leading to the
need to provide the document type all the time.

DanielSiepmann committed:
TASK: Use a single array utility

Instead of only adding a new class for one static method, we extend the
existing class to only import a single class.

This way we extend the original array utility function collection.

DanielSiepmann committed:
TASK: Remove unused code

E.g. use statement and injected object manager.

DanielSiepmann committed:
TASK: Remove wrong phpdoc

DanielSiepmann committed:
TASK: Do not publish internal method

DanielSiepmann committed:
TASK: Remove cache implementation

As it does not use TYPO3 caching framework but custom feature.

DanielSiepmann committed:
FEATURE: Provide basic caching

DanielSiepmann commented:
@bmoex I've updated the code and documentation.
I've also tested the current state with one of our projects.

Would you like to test my changes with your project and provide feedback? There were some changes that might need updates in your project.

DanielSiepmann committed:
FEATURE: Provide more information about concept
